core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -191,44 +191,219 @@ macro_rules! cast_int_to_int_macro {
             .as_any()
             .downcast_ref::<PrimitiveArray<$from_arrow_primitive_type>>()
             .unwrap();
+
         let spark_int_literal_suffix = match $from_data_type {
             &DataType::Int64 => "L",
             &DataType::Int16 => "S",
             &DataType::Int8 => "T",
             _ => "",
         };
 
-        let output_array = match $eval_mode {
-            EvalMode::Legacy => cast_array
-                .iter()
-                .map(|value| match value {
-                    Some(value) => {
-                        Ok::<Option<$to_native_type>, CometError>(Some(value as $to_native_type))
-                    }
-                    _ => Ok(None),
-                })
-                .collect::<Result<PrimitiveArray<$to_arrow_primitive_type>, _>>(),
-            _ => cast_array
-                .iter()
-                .map(|value| match value {
-                    Some(value) => {
+        let output_array = cast_array
```

Comment:
Unrelated: Made function a little bit more idiomatic and used new cast_overflow() function.

Reply:
Reverted this due to perf regression.